### PR TITLE
Add fp16 support to fused batchnorm op

### PR DESCRIPTION
**Pull request by nluehr**
Attention @zheng-xq 

- This commit adds a mixed-precision fused_batch_norm_v2 op.
  The inputs and outputs are fp16, while the scale, offset, mean
  and variance are kept in fp32.
- The tf.nn.fused_batch_norm op has been modified to use the v2
  fused batchnorm whenever inputs are fp16 (this does not affect
  compatibility because fp16 was not previously supported).
- The high-level layers API has also been updated to store the
  scale, offset, mean and variance variables as fp32.

### CODE REVIEW

**Comment by nluehr** [comment]
I signed it.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -319,9 +319,10 @@ def _fused_batch_norm(
                        (inputs.name, params_shape))
```
Comment (reedwm) [inline tensorflow/contrib/layers/python/layers/layers.py]: Remove the TODO(reedwm) above. Github doesn't let me comment on a line far away from a change. (Any idea if doing so is possible?)

File: tensorflow/python/layers/normalization.py [hunk 2]
```diff
@@ -248,6 +263,7 @@ def build(self, input_shape):
         def _renorm_variable(name, shape):
           var = self.add_variable(name=name,
                                   shape=shape,
+                                  dtype=param_dtype,
                                   initializer=init_ops.zeros_initializer(),
                                   trainable=False)
           return var
```
Comment (reedwm) [inline tensorflow/python/layers/normalization.py]: In the function "batch_normalization" below, when calling "BatchNormalization", pass the dtype of the input tensor.

Comment (reedwm) [inline tensorflow/python/layers/normalization.py]: Remove the TODO(reedwm) below.

**Review by reedwm — CHANGES_REQUESTED** [review]
Also add tests to normalization_test.py, layers_test.py, and CPU tests to nn_fused_batchnorm_test.py

File: tensorflow/contrib/layers/python/layers/layers.py (hunk 1 above)
Comment (nluehr) [reply tensorflow/contrib/layers/python/layers/layers.py]: Done

File: tensorflow/python/layers/normalization.py (hunk 2 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization.py]: Done

Comment (nluehr) [reply tensorflow/python/layers/normalization.py]: Done

File: tensorflow/python/layers/normalization.py [hunk 3]
```diff
@@ -692,6 +709,7 @@ def batch_normalization(inputs,
       trainable=trainable,
       virtual_batch_size=virtual_batch_size,
       name=name,
+      dtype=inputs.dtype,
```
Comment (reedwm) [inline tensorflow/python/layers/normalization.py]: Change "inputs.dtype" to "inputs.dtype.base_dtype"

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/layers/normalization.py (hunk 3 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization.py]: Done.

File: tensorflow/core/kernels/fused_batch_norm_op.cc [hunk 4]
```diff
@@ -54,25 +54,20 @@ struct FusedBatchNorm<CPUDevice, T, U> {
                   Tensor* batch_var_output, Tensor* saved_mean_output,
```
Comment (reedwm) [inline tensorflow/core/kernels/fused_batch_norm_op.cc]: In case you missed it before: Also add tests to normalization_test.py, layers_test.py, and CPU tests to nn_fused_batchnorm_test.py.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/fused_batch_norm_op.cc (hunk 4 above)
Comment (nluehr) [reply tensorflow/core/kernels/fused_batch_norm_op.cc]: Sorry, I had missed it. Tests included in latest commit.

File: tensorflow/python/layers/normalization_test.py [hunk 5]
```diff
@@ -90,15 +93,73 @@ def _train(self,
...
+                                                 train1_use_gpu=True,
+                                                 train2_use_gpu=True,
+                                                 infer_use_gpu=True)
+ 
+    self.assertEqual(len(ref_vars), 5)
+
+    for train1_use_gpu in [True, False]:
+      for train2_use_gpu in [True, False]:
+        for infer_use_gpu in [True, False]:
+          v, l = self._trainEvalSequence(dtypes.float16,
```
Comment (reedwm) [inline tensorflow/python/layers/normalization_test.py]: l is a hard to read variable name. Name it something like "loss" instead.

File: tensorflow/python/layers/normalization_test.py [hunk 6]
```diff
@@ -90,15 +93,73 @@ def _train(self,
...
+    train_vars = self._train(
+        checkpoint,
+        shape,
+        use_gpu=train2_use_gpu,
+        is_fused=True,
+        restore=True,
+        freeze_mode=False,
+        dtype=dtype)
+
+    image_val = np.random.rand(batch,
```
Comment (reedwm) [inline tensorflow/python/layers/normalization_test.py]: Call np.random.seed(0) right before this line, to make it obvious this this is the same value every time.

File: tensorflow/python/layers/normalization_test.py [hunk 7]
```diff
@@ -218,6 +279,34 @@ def testCreateBN(self):
...
+                                       dtype=dtypes.float16)
+    training = array_ops.placeholder(dtype='bool')
+    outputs = bn.apply(inputs, training=training)
+
+    # Verify shape.
+    self.assertListEqual(outputs.get_shape().as_list(), [5, 4, 3, 3])
+
+    # Verify layer attributes.
+    self.assertEqual(len(bn.updates), 2)
+    self.assertEqual(len(bn.variables), 4)
```
Comment (reedwm) [inline tensorflow/python/layers/normalization_test.py]: Also test the variables are float32.

File: tensorflow/python/ops/nn_fused_batchnorm_test.py [hunk 8]
```diff
@@ -341,16 +341,16 @@ def testInference(self):
             x_shape, dtype, [1], np.float32, use_gpu=True, data_format='NHWC')
         self._test_inference(
             x_shape, dtype, [1], np.float32, use_gpu=True, data_format='NCHW')
-    self._test_inference(
-        x_shape, np.float32, [1], np.float32, use_gpu=False, data_format='NHWC')
+        self._test_inference(
```
Comment (reedwm) [inline tensorflow/python/ops/nn_fused_batchnorm_test.py]: This test is removed in the case where no GPU is available, and same for the test below. Like you did for the tests below, swap the "if test.is_gpu_available(cuda_only=True)" and "for dtype in [np.float16, np.float32]" statements

File: tensorflow/python/layers/normalization_test.py [hunk 9]
```diff
@@ -68,19 +68,22 @@ def _train(self,
              use_gpu,
              is_fused,
              restore=False,
-             freeze_mode=False):
+             freeze_mode=False,
+             dtype=None):
```
Comment (reedwm) [inline tensorflow/python/layers/normalization_test.py]: Have dtype default to dtypes.float32. Then you don't have to have the "if dtype is None" check below.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_fused_batchnorm_test.py (hunk 8 above)
Comment (nluehr) [reply tensorflow/python/ops/nn_fused_batchnorm_test.py]: Done

File: tensorflow/python/layers/normalization_test.py (hunk 7 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization_test.py]: Done.

File: tensorflow/python/layers/normalization_test.py (hunk 5 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization_test.py]: Done.

File: tensorflow/python/layers/normalization_test.py (hunk 6 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization_test.py]: Done.

File: tensorflow/python/layers/normalization_test.py (hunk 9 above)
Comment (nluehr) [reply tensorflow/python/layers/normalization_test.py]: Done.

**Review by reedwm — APPROVED** [review]

**Comment by gunan** [comment]
Looks like some sanity checks failed

FAIL: Found 2 non-whitelited pylint errors:
tensorflow/python/layers/normalization.py:246: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 6

tensorflow/python/layers/normalization.py:248: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 6

**Comment by nluehr** [comment]
Fixed.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
CLA bot seems to be confused with multiple commits again. As all the CLAs seem to be signed, I can merge once the tests have passed.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Looks like api compatibility test has detected a change in the public TF python API.
```
ERROR:tensorflow:1 differences found between API and golden.
ERROR:tensorflow:Issue 1	: Change detected in python object: tensorflow.layers.BatchNormalization.
```

Could you confirm the change is intended, run the test in a python2 + linux setup (you may use our devel docker images) and pack the change into this pull request?
Then we will need an API review before we can merge this PR.

**Comment by nluehr** [comment]
The API change was actually unnecessary, as dtype can be passed to the underlying base.Layer in kwargs. The current PR should pass the api compatibility test.

**Comment by gunan** [comment]
Thanks for checking. Could you also look into the merge conflicts before I trigger tests?

**Comment by nluehr** [comment]
The conflict has been resolved.

**Comment by gunan** [comment]
Jenkins, test this please.
